lib/rspec/expectations/block_snippet_extractor.rb

```diff
@@ -0,0 +1,241 @@
+        attr_reader :state, :body_tokens
+
+        def initialize(*)
+          super
+          parse!
+        end
+
+        private
+
+        def parse!
+          @state = :initial
+
+          catch(:finish) do
+            source.tokens.each do |token|
+              invoke_state_handler(token)
+            end
+            raise TargetNotFoundError, 'Could not find the target block'
+          end
+        end
+
+        def finish!
+          throw :finish
+        end
+
+        def invoke_state_handler(token)
+          __send__("#{state}_state", token)
+        end
+
+        def initial_state(token)
+          @state = :after_method_call if token.location == block_locator.method_call_location
+        end
+
+        def after_method_call_state(token)
+          @state = :after_opener if handle_opener_token(token)
+        end
+
+        def after_opener_state(token)
+          if handle_closer_token(token)
+            finish_or_find_next_block_if_incorrect!
+          elsif pipe_token?(token)
+            finalize_pending_tokens!
+            @state = :after_beginning_of_args
+          else
+            pending_tokens << token
+            handle_opener_token(token)
+            @state = :after_beginning_of_body unless token.sp_type?
+          end
+        end
+
+        def after_beginning_of_args_state(token)
+          @state = :after_beginning_of_body if pipe_token?(token)
+        end
+
+        def after_beginning_of_body_state(token)
+          if handle_closer_token(token)
+            finish_or_find_next_block_if_incorrect!
+          else
+            pending_tokens << token
+            handle_opener_token(token)
+          end
+        end
+
+        def pending_tokens
+          @pending_tokens ||= []
+        end
+
+        def finalize_pending_tokens!
+          pending_tokens.freeze.tap do
+            @pending_tokens = nil
+          end
+        end
+
+        def finish_or_find_next_block_if_incorrect!
+          body_tokens = finalize_pending_tokens!
+
+          if correct_block?(body_tokens)
+            @body_tokens = body_tokens
+            finish!
+          else
+            @state = :after_method_call
+          end
+        end
+
+        def handle_opener_token(token)
+          opener_token?(token).tap do |boolean|
+            opener_token_stack.push(token) if boolean
+          end
+        end
+
+        def opener_token?(token)
+          token.lbrace_type? || (token.kw_type? && token.string == 'do')
+        end
+
+        def handle_closer_token(token)
+          if opener_token_stack.last.closed_by?(token)
+            opener_token_stack.pop
+            opener_token_stack.empty?
+          else
+            false
+          end
+        end
+
+        def opener_token_stack
+          @opener_token_stack ||= []
+        end
+
+        def pipe_token?(token)
+          token.op_type? && token.string == '|'
+        end
+
+        def correct_block?(body_tokens)
```

Comment:
When parsing the following code:

```ruby
target_method(another_method { 2.negative? }) { 1.positive? }
#                                          ^
```

... BlockTokenExtractor iterates the following tokens:

```ruby
[
  [[1, 0], :on_ident, "target_method"],
  [[1, 13], :on_lparen, "("],
  [[1, 14], :on_ident, "another_method"],
  [[1, 28], :on_sp, " "],
  [[1, 29], :on_lbrace, "{"],
  [[1, 30], :on_sp, " "],
  [[1, 31], :on_int, "2"],
  [[1, 32], :on_period, "."],
  [[1, 33], :on_ident, "negative?"],
  [[1, 42], :on_sp, " "],
  [[1, 43], :on_rbrace, "}"], # <
  [[1, 44], :on_rparen, ")"],
  [[1, 45], :on_sp, " "],
  [[1, 46], :on_lbrace, "{"],
  [[1, 47], :on_sp, " "],
  [[1, 48], :on_int, "1"],
  [[1, 49], :on_period, "."],
  [[1, 50], :on_ident, "positive?"],
  [[1, 59], :on_sp, " "],
  [[1, 60], :on_rbrace, "}"]
]
```

Because there's no semantic information in tokens, BlockTokenExtractor does not know which block is the correct target. So, at the `^` point, BlockTokenExtractor asks BlockLocator (which utilizes semantic AST information) if the last parsed block (`{ 2.negative? }`) is the correct target or not. If it's incorrect, it continues further parsing.